File: services/jgrapht_service/src/process/src/main/java/process/algorithm/Algorithms.java
package process.algorithm;

import lombok.RequiredArgsConstructor;
import process.algorithm.coloring.ColoringBacktrack;
import process.algorithm.coloring.ColoringGeneticAlgorithm;
import process.algorithm.coloring.TabuCol;
import process.algorithm.shortestpath.AStar;
import process.algorithm.shortestpath.BellmanFord;
import process.algorithm.shortestpath.Dijkstra;
import process.instance.Instance;
import process.pubsub.Publisher;

import java.util.List;

@RequiredArgsConstructor
public class Algorithms {

    private final Instance instance;
    private final List<Publisher> publishers;

    public Algorithm getAlgorithm(String algorithmName) throws Exception {
        return switch (algorithmName) {
            case "Dijkstra" -> new Dijkstra(instance, publishers);
            case "Bellman-Ford" -> new BellmanFord(instance, publishers);
            case "A*" -> new AStar(instance, publishers);
            case "Coloring Backtrack" -> new ColoringBacktrack(instance, publishers);
            case "TabuCol" -> new TabuCol(instance, publishers);
            case "Coloring Genetic Algorithm" -> new ColoringGeneticAlgorithm(instance, publishers);
            default -> throw new Exception("Algorithm not implemented");
        };
    }
}

File: services/networkx_service/src/process/algorithm/algorithms.py
from .shortest_path.astar import AStar
from .shortest_path.bellman_ford import BellmanFord
from .shortest_path.dijkstra import Dijkstra
from .coloring.coloring_backtrack import ColoringBacktrack
from .coloring.tabucol import TabuCol
from .coloring.coloring_genetic_algorithm import ColoringGeneticAlgorithm


class Algorithms:
    def __init__(self, instance, publishers):
        self.instance = instance
        self.publishers = publishers

    def get_algorithm(self, algorithm_name):
        match algorithm_name:
            case "Dijkstra":
                return Dijkstra(self.instance, self.publishers)
            case "Bellman-Ford":
                return BellmanFord(self.instance, self.publishers)
            case "A*":
                return AStar(self.instance, self.publishers)
            case "Coloring Backtrack":
                return ColoringBacktrack(self.instance, self.publishers)
            case "TabuCol":
                return TabuCol(self.instance, self.publishers)
            case "Coloring Genetic Algorithm":
                return ColoringGeneticAlgorithm(self.instance, self.publishers)
            case _:
                raise NotImplementedError("Algorithm not implemented")
